tls/s2n_x509_validator.c

```diff
@@ -0,0 +1,430 @@
+                            uint8_t check_ocsp, verify_host verify_host_fn, void *verify_ctx) {
+    notnull_check(trust_store);
+    validator->trust_store = trust_store;
+
+    validator->validate_certificates = 1;
+    validator->check_stapled_ocsp = check_ocsp;
+
+    validator->verify_host_fn = default_verify_host;
+
+    if (verify_host_fn) {
+        validator->verify_host_fn = verify_host_fn;
+    }
+
+    validator->validation_ctx = verify_ctx;
+
+    validator->cert_chain = NULL;
+    if (validator->trust_store->trust_store) {
+        validator->cert_chain = sk_X509_new_null();
+    }
+
+    return 0;
+}
+
+void s2n_x509_validator_cleanup(struct s2n_x509_validator *validator) {
+    if (validator->cert_chain) {
+        X509 *cert = NULL;
+        while ((cert = sk_X509_pop(validator->cert_chain))) {
+            X509_free(cert);
+        }
+
+        sk_X509_free(validator->cert_chain);
+    }
+
+    validator->trust_store = NULL;
+    validator->verify_host_fn = NULL;
+    validator->validation_ctx = NULL;
+    validator->validate_certificates = 0;
+}
+
+//for each name in the cert. Iterate them. Call the callback. If one returns true, then consider it validated,
+//if none of them return true, the cert is considered invalid.
+static uint8_t verify_host_information(struct s2n_x509_validator *validator, X509 *public_cert) {
+    uint8_t verified = 0;
+
+    //get each subject alternative name.
+    STACK_OF(GENERAL_NAME) *names_list = X509_get_ext_d2i(public_cert, NID_subject_alt_name, NULL, NULL);
+    GENERAL_NAME *current_name = NULL;
+    while (!verified && names_list && (current_name = sk_GENERAL_NAME_pop(names_list))) {
+        const char *name = (const char *) M_ASN1_STRING_data(current_name->d.ia5);
+        size_t name_len = (size_t) M_ASN1_STRING_length(current_name->d.ia5);
+
+        verified = validator->verify_host_fn(name, name_len, validator->validation_ctx);
+    }
+
+    GENERAL_NAMES_free(names_list);
+
+    //if none of those were valid, go to the subject name.
+    if (!verified) {
+        X509_NAME *subject_name = X509_get_subject_name(public_cert);
+        if (subject_name) {
+            int j = 0, i = -1;
+            while ((j = X509_NAME_get_index_by_NID(subject_name, NID_commonName, i)) >= 0) {
+                i = j;
+            }
+
+            if (i >= 0) {
+                ASN1_STRING *common_name =
+                        X509_NAME_ENTRY_get_data(X509_NAME_get_entry(subject_name, i));
+
+                if (common_name) {
+                    char peer_cn[255];
+                    static size_t peer_cn_size = sizeof(peer_cn);
+                    memset_check(&peer_cn, 0, peer_cn_size);
+                    if (ASN1_STRING_type(common_name) == V_ASN1_UTF8STRING) {
+                        size_t len = (size_t) ASN1_STRING_length(common_name);
+                        //save space for the null terminator
+                        len = len > sizeof(peer_cn) - 1 ? sizeof(peer_cn) - 1 : len;
```

Comment:
This line scares me. Are we truncating the name being validated if it exceeds 255 characters? 

Isn't it possible for `example.com.evil.com` to get truncated to `example.com` if it was actually 255 characters long? We should instead do `lte_check(len, sizeof(peer_cn) - 1);` and fail the verification if anything goes out of bounds.

Reply:
Good callout. anything over 254 characters is illegal per the spec if IIRC anyways (which is why I chose 255).